app/Http/Controllers/CouponController.php

```diff
@@ -0,0 +1,223 @@
+<?php
+
+namespace App\Http\Controllers;
+use App\Http\Interfaces\CrudInterface_FH;
+use App\Http\Interfaces\PaymentServiceInterface;
+use App\Http\Requests\CouponRequest;
+use App\Http\Resources\CouponResource;
+use App\Models\Coupon;
+use App\Models\CouponCode;
+use App\Models\Product;
+use Illuminate\Support\Str;
+
+class CouponController extends Controller implements CrudInterface_FH
+{
+    protected $paymentService;
+    private $defaultPageSize = 10;
+
+    // Injected Service 
+    public function __construct(PaymentServiceInterface $paymentService)
+    {
+        $this->paymentService = $paymentService;
+    }
+// -- to fetch all coupons with pagination --
+    public function index() {
+        try {
+           $coupons = Coupon::paginate($this->defaultPageSize);
+           return successResponse("Coupon fetched successfully!", CouponResource::collection($coupons),true, 200);
+        } catch (\Exception $e) {
+            return handleException($e);
+        }
+    }
+
+// -- get a specific coupon by id --
+     public function show($id) {
+        try {
+            $coupon = Coupon::with("couponCodes")->findOrFail($id);
+            return successResponse("Coupon fetched successfully!", CouponResource::make($coupon));
+        }
+        catch (\Exception $e) {
+            return handleException($e);
+        }
+     }
+
+// -- Create a coupon with coupon codes --
+     public function create(CouponRequest $request)
+     {
+         $validatedData = $request->validated();
+         return $this->store($validatedData);
+     }
+
+ 
+     public function store(array $payload) {
+        try {
+            // Check if coupon is for product or cart
+            if (array_key_exists("product_id" ,$payload) && !is_null($payload['product_id'])) {
+              // ----- Coupon is for product -----
+                //get product price
+                $product = Product::find($payload["product_id"]);
```

Comment:
what if product not found the code will crash on the very next line as you are using price from result set of product

Reply:
use findOrfail here